hack/typelinter/typelinter.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/types"
+	"os"
+	"regexp"
+
+	"golang.org/x/tools/go/packages"
+)
+
+type stringListVar map[string]struct{}
+
+func (sl stringListVar) String() string {
+	var keys []string
+	for k := range sl {
+		keys = append(keys, k)
+	}
+	return fmt.Sprintf("%v", keys)
+}
+
+func (sl *stringListVar) Set(value string) error {
+	if *sl == nil {
+		*sl = map[string]struct{}{}
+	}
+	(*sl)[value] = struct{}{}
+	return nil
+}
+
+var (
+	cursedFields   stringListVar
+	typesWhitelist string
+
+	typesWhitelistRegexp *regexp.Regexp
+)
+
+func validStruct(structName string, s *types.Struct) bool {
+	valid := true
+	for i := 0; i < s.NumFields(); i++ {
+		field := s.Field(i)
+		fieldName := fmt.Sprintf("%s:%s", structName, field.Name())
+		if _, ok := cursedFields[fieldName]; ok {
+			continue
+		}
+		typ := field.Type().String()
+		if !typesWhitelistRegexp.MatchString(typ) {
+			fmt.Fprintf(os.Stderr, "%s: type %s is not allowed to be used\n", fieldName, typ)
+			valid = false
+		}
+	}
+	return valid
+}
+
+func validPackage(pkg *packages.Package) bool {
+	valid := true
+	scope := pkg.Types.Scope()
+	for _, name := range scope.Names() {
+		obj := scope.Lookup(name)
+		if !obj.Exported() {
+			continue
+		}
+		if typeName, ok := obj.(*types.TypeName); ok {
+			typ := typeName.Type().Underlying()
+			if s, ok := typ.(*types.Struct); ok {
+				structName := fmt.Sprintf("%s.%s", pkg.PkgPath, typeName.Name())
+				if !validStruct(structName, s) {
+					valid = false
+				}
+			}
+		}
+	}
+	return valid
+}
+
+func main() {
+	flag.Var(&cursedFields, "cursed", "exclude the field from being checked (e.g. -cursed=github.com/openshift/api/image/dockerpre012.ImagePre012:Created), can be used multiple times")
```

Comment:
excluded

Reply:
it's the imperative mood: `exclude the field ...`
would you prefer `the field excluded from being checked`?

Reply:
updated